### ISSUE TITLE
请问MI-FGSM的实现中,并未考虑对初始样本的梯度,因为梯度为0吗

### ISSUE THREAD

**Issue by QLJiao** [BorealisAI/advertorch#15]

**Comment by gwding** [comment]
I guess you're asking about `LinfBasicIterativeAttack` at https://github.com/BorealisAI/advertorch/blob/master/advertorch/attacks/iterative_projected_gradient.py#L235

The first step of the attack is indeed based on the gradient of the original example. It is calculated as the delta's gradient when delta is all zeros.

**Comment by QLJiao** [comment]
@gwding no,it's about MomentumIterativeAttack, the gradient of the original image is not included

def perturb(self, x, y=None):
        """
        Given examples (x, y), returns their adversarial counterparts with
        an attack length of eps.
        :param x: input tensor.
        :param y: label tensor.
                  - if None and self.targeted=False, compute y as predicted
                    labels.
                  - if self.targeted=True, then y must be the targeted labels.
        :return: tensor containing perturbed inputs.
        """
        x, y = self._verify_and_process_inputs(x, y)

        delta = torch.zeros_like(x)
        g = torch.zeros_like(x)

        delta = nn.Parameter(delta)

        for i in range(self.nb_iter):

            if delta.grad is not None:
                delta.grad.detach_()
                delta.grad.zero_()

            imgadv = x + delta
            outputs = self.predict(imgadv)
            loss = self.loss_fn(outputs, y)
            if self.targeted:
                loss = -loss
            loss.backward()

            g = self.decay_factor * g + normalize_by_pnorm(
                delta.grad.data, p=1)

**Comment by gwding** [comment]
@QLJiao sorry about my misunderstanding, of course it is `MomentumIterativeAttack`. 

The answer would be similar, if you look at this chunk of code https://github.com/BorealisAI/advertorch/blob/master/advertorch/attacks/iterative_projected_gradient.py#L317-L353

when `i==0`, `delta` is zero, so `delta.grad` is the same as "the gradient of the original image", does that make sense?

**Comment by gwding** [comment]
@QLJiao I hope my comments were helpful. I'm closing this issue for now, please reopen if it is still confusing.

**Comment by rshaojimmy** [comment]
> @QLJiao sorry about my misunderstanding, of course it is `MomentumIterativeAttack`.
> 
> The answer would be similar, if you look at this chunk of code https://github.com/BorealisAI/advertorch/blob/master/advertorch/attacks/iterative_projected_gradient.py#L317-L353
> 
> when `i==0`, `delta` is zero, so `delta.grad` is the same as "the gradient of the original image", does that make sense?

May I further figure out why 'delta.grad' is the same as "the gradient of the original image" when 'delta' is zero?
